feat: add dark and light theme switch support

svnset opened the pull request:
Added another build file to build theme variants consisting of the light theme latte plus the `dark_theme` variants.

The variants are safely written to `themes/dark-light/.`

The `dark_theme` manifest property respects the system setting of preferring dark or light themes and switches the theme immediately once either option is changed.

See [here](https://webextension-api.thunderbird.net/en/beta-mv2/theme.html#manifest-file-properties) for reference.

Finally I think it's a great idea to have a separate build option here, but to not mix the two, as there are a few disadvantages as well, eg.of having to rely on the `prefer-dark` value, or else being forced to use the latte theme.

Commit:
feat: add dark and light theme switch support

Update build file to support both a theme with and without dark_theme switch support.
Update catppuccin/palette objects using newer deno library.

uncenter commented:
Why are we basically duplicating much of the work of the previously existing build script here? This should be added to `build.ts` and reuse the settings and other things from it.

svnset commented:
Read my last sentence, there are a couple of disadvantages. You will rely on `prefer_dark`and `prefer_light` system values, if you do not want that feature you are stuck with the latte version. We could just generate both versions in 1 script if that's what you mean. Also what do you mean with reusing settings?  This script does reuse everything that was offered in `build.ts` but it was also updated to use  the newest catppuccin/palette version and objects.

uncenter commented:
> We could just generate both versions in 1 script if that's what you mean.
> This script does reuse everything that was offered in `build.ts` but it was also updated to use the newest catppuccin/palette version and objects.

Yes. For example you duplicated the accent array near the top, I presume the settings are the same for the most part but just nested under different objects, etc. It would make more sense to me to add this new capability/feature to the existing build script.

svnset commented:
I think I get what you mean, but if you diff the 2 files you see that there is not much difference at all. I just leverage the new palette objects and I did not touch the theming at all. If we decide to combine it with the old build.ts I would still rewrite it to be up to date with `catppuccin/palette`. 

So you are suggesting we still create 2 versions but generate these into 2 different folders with 1 `build.ts` script doing both. For now this would result in a very non-pretty build file as there is not really a good way to do both themes within 1 iteration. I will give it a go and hopefully come up with something bearable.

svnset commented:
Here is a first version of how a build file offering both variants could look like.
The theme itself is not touched, i just duplicated the darker variants in to the `dark_theme`, it looks very good and swaps instantly as soon as you change the system setting dark/light. The new theme versions are written to `./themes/dark-light/.` but we can change that of course.
 
I updated the `catpuccin/palette` to `https://deno.land/x/catppuccin@v1.7.1/mod.ts`and updated the objects to their equivalents.

uncenter commented:
No need to force push, we use squash merging - it's actually preferred to use individual commits because force pushes are harder to review on GitHub.

uncenter commented on build.ts:
```diff
@@ -1,10 +1,9 @@
+  CatppuccinFlavor,
+  flavorEntries,
+  flavors,
+} from "https://deno.land/x/catppuccin@v1.7.1/mod.ts";
```
```suggestion
} from "https://deno.land/x/catppuccin@v1.7.1/mod.ts";
```
Let's use npm/JSR, deno.land/x is basically abandoned/deprecated at this point.

uncenter commented on build.ts:
```diff
@@ -32,11 +31,18 @@ const accents = [
   name: string,
-  accent: keyof Labels<Color, AlphaColor>,
-  palette: Labels<Color, AlphaColor>
+  accent: string,
```
```suggestion
  accent: AccentName,
```
You can use the `AccentName` type from `@catppuccin/palette`.

uncenter commented on build.ts:
```diff
@@ -32,11 +31,18 @@ const accents = [
+      themeAccent = hex;
+    }
+  });
+  const palette = flavor.colors;
```
I don't know why we really need to change from `palette[accent].hex` since it seems more clean and clear to me, but if we do this map and everything is totally unnecessary.
```suggestion
  const palette = flavor.colors;
  const themeAccent = palette[accent].hex;
```

uncenter commented on build.ts:
```diff
@@ -47,9 +53,9 @@ async function makeThemeObject(
-      name
-    )} ${titleCase(accent)}`,
+    description: `Soothing pastel theme for Thunderbird - ${titleCase(name)} ${
+      titleCase(accent)
```
```suggestion
      palette[accent].name
```

uncenter replied:
This isn't resolved? You didn't change this. We should use the `name` property of the accent color and flavor instead of this `titleCase `function.

uncenter commented on build.ts:
```diff
@@ -144,33 +342,44 @@ async function makeThemeObject(
-  Object.entries(variants).map(([name, palette]) =>
-    generateVariants(name, palette)
-  )
+  flavorEntries.map(([name, flavor]) => generateVariants(name, flavor)),
```
```suggestion
  flavorEntries.map(([identifier, flavor]) => generateVariants(identifier, flavor)),
```
Catppuccin's palette metadata has distinct differences between "identifiers" and "names" (e.g. `frappe` is the identifier whereas `Frappé` is the name), it helps to keep it consistent in variable naming as well.

svnset replied:
This one I still could do. I guess the Name could be used in the title and the identifier everywhere else.

svnset replied:
done :)

uncenter commented on build.ts:
```diff
@@ -144,33 +342,44 @@ async function makeThemeObject(
+
+async function writeTheme(
+  fileName: string,
+  theme: unknown,
```
```suggestion
  theme: unknown,
```
Let's make and use a generic `ThemeObject` record type instead of saying this is unknown.

svnset replied:
I created a ThemeObject to be more type safe.

uncenter commented on build.ts:
```diff
@@ -144,33 +342,44 @@ async function makeThemeObject(
+  theme: unknown,
+  path: string,
+) {
+  if (theme) {
```
```suggestion
```
I think this is unnecessary with proper typing.

svnset replied:
It's the most easy check to avoid the double latte theme case.

uncenter replied:
No it's not. It's more confusing to the control loop because then you have to go and track down why `theme` can be undefined. We should just not try to make a theme object for latte to begin with and save ourselves the trouble.

svnset replied:
I changed it.

uncenter commented on build.ts:
```diff
@@ -144,33 +342,44 @@ async function makeThemeObject(
 
-    await zip.writeZip(`./themes/${name}/${name}-${accent}.xpi`);
+    await zip.writeZip(`${path}/${fileName}`);
   }
```
```suggestion
```
See above.

uncenter commented on build.ts:
```diff
@@ -84,43 +92,233 @@ async function makeThemeObject(
+        lightAccent = hex;
+      }
+    });
+  } else return NaN;
```
Why are we using `NaN` as the invalid return type? That doesn't make sense since this function is supposed to return an object. Use an empty return (undefined) or return `null`.

svnset replied:
I mean you are correct and not as I just do a if (theme) check later and that checks for NaN, null etc but I will change it sure or maybe we do omit in the outer loop like you suggested in the other comment.

uncenter commented on build.ts:
```diff
@@ -84,43 +92,233 @@ async function makeThemeObject(
+  accent: string,
+  flavor: CatppuccinFlavor,
+) {
+  let lightAccent: string = "";
+  let darkAccent: string = "";
+
+  if (flavor.dark) {
+    flavor.colorEntries.map(([colorName, { hex }]) => {
+      if (colorName == accent) {
+        darkAccent = hex;
+      }
+    });
+    flavors.latte.colorEntries.map(([colorName, { hex }]) => {
+      if (colorName == accent) {
+        lightAccent = hex;
+      }
+    });
+  } else return NaN;
```
Since this function isn't meant to operate on the latte flavor, we could add a clause to skip calling it for latte in the main for loop, and use `Omit` or something to make that clear here.
```suggestion
  let lightAccent: flavor.colors[accent].hex;
  let darkAccent: flavors.latte.colors[accent].hex;
```

svnset replied:
I did not put too much thought into this and worked with what `flavor` had to offer like the `dark` boolean but this could be done better without the maps agreed.

uncenter commented on build.ts:
```diff
@@ -84,43 +92,233 @@ async function makeThemeObject(
+    dark_theme: {
+      colors: {
+        frame: palette.base.hex,
+        button_background_active: darkAccent,
+        button_background_hover: palette.surface0.hex,
+        icons: palette.text.hex,
+        tab_text: palette.text.hex,
+        tab_line: darkAccent,
+        tab_loading: darkAccent,
+        tab_selected: palette.base.hex,
+        tab_background_text: palette.overlay1.hex,
+        tab_background_separator: palette.surface0.hex,
+        bookmark_text: palette.text.hex,
+        toolbar: palette.base.hex,
+        toolbar_field: palette.surface0.hex,
+        toolbar_field_text: palette.text.hex,
+        toolbar_field_highlight: darkAccent,
+        toolbar_field_highlight_text: palette.mantle.hex,
+        toolbar_field_border: palette.mantle.hex,
+        toolbar_field_focus: palette.surface0.hex,
+        toolbar_field_text_focus: palette.text.hex,
+        toolbar_field_border_focus: darkAccent,
+        toolbar_top_separator: palette.surface0.hex,
+        toolbar_bottom_separator: palette.surface0.hex,
+        toolbar_vertical_separator: palette.surface0.hex,
+        sidebar: palette.base.hex,
+        sidebar_text: palette.text.hex,
+        sidebar_highlight: darkAccent,
+        sidebar_highlight_text: palette.mantle.hex,
+        sidebar_border: palette.surface0.hex,
+        popup: palette.surface0.hex,
+        popup_text: palette.text.hex,
+        popup_border: palette.base.hex,
+        popup_highlight: darkAccent,
+        popup_highlight_text: palette.mantle.hex,
+        spaces_bg: palette.mantle.hex,
+        tree_view_bg: palette.crust.hex,
+        bg_color: palette.base.hex,
+        spaces_bg_active: darkAccent,
+        button_primary_bg: darkAccent,
```
A lot of this is duplicated from the other build function. We should have a function to generate the `theme.colors` from a set of colors so we don't duplicate the settings for the main function, the dark section, and the light section?

svnset replied:
I think a function to return theme.colors is a good idea to no repeat the theme too often. I will work on that.

svnset commented:
Hello again, thank you for taking the time to review and all the suggestions. The force pushes were only minor things like newline, formatting and import order that's why I force pushed amend to not spam with unnecessary commits :)

I like your suggestions and will see to incorporate everything in.

Commit:
feat: dark light theme support

added a more elegant solution to multiply the theme from one common
theme that has alwas `theme` and `dark_theme` and uses the 3 color
scheme properties dark light and auto.

svnset commented:
Okay so I did put some more effort into this. I found a good way of leveraging one common theme to generate both variants. We can always use dark_theme and just differentiate between the different modes with the `color_scheme` property. I tried to incorporate all of your suggestions aside from making a simple null check for the latte + latte case. 

Please have a look :)

Commit:
changed name to identifier and used proper name in theme title

uncenter commented on build.ts:
```diff
@@ -71,106 +88,135 @@ async function makeThemeObject(
   };
 }
 
+async function getColors(
```
This doesn't need to be async?

svnset replied:
Nope it can be non async.

uncenter commented on build.ts:
```diff
@@ -31,12 +31,29 @@ const accents = [
-  palette: Labels<Color, AlphaColor>
+  identifier: string,
+  accent: AccentName,
+  palette: CatppuccinFlavor,
```
We usually refer to `CatppuccinFlavor['colors']` as `palette`, this should just be called `flavor`.
```suggestion
  flavor: CatppuccinFlavor,
```

uncenter commented on build.ts:
```diff
@@ -31,12 +31,29 @@ const accents = [
-  const themeName = `catppuccin-${name}-${accent}`;
-  const encodedName = new TextEncoder().encode(themeName)
+  const darkColors = await getColors(accent, palette);
+  let lightColors = await getColors(accent, palette);
```
```suggestion
  const darkColors = await getColors(accent, flavor.colors);
  let lightColors = await getColors(accent, flavor.colors);
```

Commit:
renamed palette to flavor; removed unnecessary async functions

Commit:
Ensured no autoswitch theme generated when flavor is latte

Changed accent to proper name in title

uncenter commented on build.ts:
```diff
@@ -195,28 +193,28 @@ async function writeTheme(
 const start = performance.now();
 await Promise.all(
-  flavorEntries.map(([identifier, flavor]) => generateVariants(identifier, flavor)),
+  flavorEntries.flatMap(([identifier, flavor]) =>
+    generateVariants(identifier, flavor)
+  ),
```
Why is this a flatMap? `generateVariants` doesn't return anything as far as I can tell.

svnset replied:
Good sight, i was trying to filter latte but I decided against it. Will fix it asap.

Commit:
fix: use map instead of flatmap

Commit:
Created a ThemeObject and added clear return types

uncenter commented on build.ts:
```diff
@@ -71,106 +114,137 @@ async function makeThemeObject(
+        color_scheme: colorScheme,
       },
     },
+    dark_theme: {
+      colors: darkColors,
+      properties: {
+        color_scheme: colorScheme,
+      },
+    },
```
It seems like we now always create this `dark_theme` property/object for themes, when we didn't in the past. Does this break anything? I would assume we would conditionally add this with a spread operator or something.

svnset replied:
That is correct, I got the idea from the official thunderbird dark/light theme itself. I did extensive testing and the important part is that we always set the correct `color_scheme` property accordingly. There is no disadvantage in setting a `dark_theme` at all.

For a single flavor selected it will always be that specific flavor and works with both `prefer_dark`and `prefer_light` in place. For the auto theme selected it will switch between the `dark_theme` (frappe, macchiato and mocha) and `theme` which is always latte.

svnset replied:
Waiting for your response before resolving this.

svnset replied:
@uncenter any update here? If it's okay for you I resolve this.

uncenter commented on build.ts:
```diff
@@ -1,18 +1,47 @@
+  flavors,
+} from "jsr:@catppuccin/palette@1.7.1";
 import { JSZip } from "https://deno.land/x/jszip@0.11.0/mod.ts";
 import * as uuid from "https://deno.land/std@0.207.0/uuid/mod.ts";
```
Might as well switch to https://jsr.io/@std/uuid here if we update the other deps.

Commit:
Exchanged deno uuid dep with jsr equivalent
